### PR TITLE
Improve logging while calling core API

### PR DESCRIPTION
**Pull request by XuanWang-Amos**
### Description

We have multiple cases that users are seeing `grpc._cython.cygrpc.ExecuteBatchError: Failed "execute_batch"` in their logs, while the root causes are different, they're all raised while executing batch operations in core, this PR is to improve logging so we can have more context on what types of error is raised in core.

### Changes included
- Use core API `grpc_call_error_to_string` to get more details from errors.
- Add additional log while handling exception and calling `_send_error_status_from_server`.

### Testing
* Error in core while starting batch:
  * Before change:
```
_cython.cygrpc.ExecuteBatchError: Failed grpc_call_start_batch: 2
```

  * After change:
```
_cython.cygrpc.ExecuteBatchError: Failed grpc_call_start_batch: 2 with grpc_call_error value: b'GRPC_CALL_ERROR_NOT_ON_SERVER'
```

* Error in RPC handler with additional exception calling `_send_error_status_from_server`:
  * Before change:
```
RROR:grpc._cython.cygrpc:Unexpected [Exception] raised by servicer method [/grpc.testing.TestService/UnaryCall]
Traceback (most recent call last):
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi", line 683, in _cython.cygrpc._handle_exceptions
Exception: test Exception in rpc_coro
ERROR:grpc._cython.cygrpc:calling _send_error_status_from_server
Traceback (most recent call last):
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi", line 683, in _cython.cygrpc._handle_exceptions
Exception: test Exception in rpc_coro
ERROR:asyncio:Task exception was never retrieved
future: <Task finished name='Task-6' coro=<<coroutine without __name__>()> exception=ExecuteBatchError('test ExecuteBatchError') created at /usr/lib/python3.10/asyncio/events.py:80>
source_traceback: Object created at (most recent call last):
....<Skipped>
```

  * After change:
```
RROR:grpc._cython.cygrpc:Unexpected [Exception] raised by servicer method [/grpc.testing.TestService/UnaryCall]
Traceback (most recent call last):
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi", line 683, in _cython.cygrpc._handle_exceptions
Exception: test Exception in rpc_coro
ERROR:grpc._cython.cygrpc:calling _send_error_status_from_server
Traceback (most recent call last):
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi", line 683, in _cython.cygrpc._handle_exceptions
Exception: test Exception in rpc_coro
ERROR:grpc._cython.cygrpc:Failed sending error status from server
Traceback (most recent call last):
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi", line 683, in _cython.cygrpc._handle_exceptions
Exception: test Exception in rpc_coro

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi", line 722, in _cython.cygrpc._handle_exceptions
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/callback_common.pyx.pxi", line 176, in _send_error_status_from_server
_cython.cygrpc.ExecuteBatchError: test ExecuteBatchError
Traceback (most recent call last):
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi", line 683, in _cython.cygrpc._handle_exceptions
Exception: test Exception in rpc_coro

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi", line 722, in _cython.cygrpc._handle_exceptions
  File "src/python/grpcio/grpc/_cython/_cygrpc/aio/callback_common.pyx.pxi", line 176, in _send_error_status_from_server
_cython.cygrpc.ExecuteBatchError: test ExecuteBatchError
```

### CODE REVIEW

**Comment by gnossen** [comment]
> Passed all CI tests.

This is not sufficient. I highly doubt this code path is covered by an existing test. Can you please manually cause a core call to fail and ensure that the result is what we expect?

File: src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi
Comment (gnossen) [inline src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi:729]: Please also log the contents of the exception, including the full stack trace.

File: src/python/grpcio/grpc/_cython/_cygrpc/aio/callback_common.pyx.pxi
Comment (gnossen) [inline src/python/grpcio/grpc/_cython/_cygrpc/aio/callback_common.pyx.pxi:32]: After some digging, it looks like this `CallbackFailureHandler` actually handles the case that `grpc_event.success` is 0, not the case that we have a bad `grpc_call_error`. We'll want to revert the changes to this class.

File: src/python/grpcio/grpc/_cython/_cygrpc/aio/callback_common.pyx.pxi
Comment (gnossen) [inline src/python/grpcio/grpc/_cython/_cygrpc/aio/callback_common.pyx.pxi:106]: Please add quotes around `grpc_call_error_string` in the message here. This will aid readability.

**Review by gnossen — COMMENTED** [review]
It looks like there _are_ a couple of other places where we could be calling `grpc_call_error_to_string` on a failed `grpc_call_error`, but there doesn't seem to be any unified error handling for all of them. Let's not block this PR on those other instances.

File: src/python/grpcio/grpc/_cython/_cygrpc/aio/callback_common.pyx.pxi
Comment (XuanWang-Amos) [reply src/python/grpcio/grpc/_cython/_cygrpc/aio/callback_common.pyx.pxi:32]: You're right, I'll revert changes to `CallbackFailureHandler`.

**Comment by XuanWang-Amos** [comment]
> > Passed all CI tests.
> 
> This is not sufficient. I highly doubt this code path is covered by an existing test. Can you please manually cause a core call to fail and ensure that the result is what we expect?

Sure, I added the result to description too.

**Comment by XuanWang-Amos** [comment]
@gnossen PTAL

File: src/python/grpcio/grpc/_cython/_cygrpc/aio/callback_common.pyx.pxi
Comment (gnossen) [reply src/python/grpcio/grpc/_cython/_cygrpc/aio/callback_common.pyx.pxi:97]: Ah, this is a `bytes` object, not a `str` object. Please decode as ascii and _then_ add quotes.

**Review by gnossen — APPROVED** [review]
LGTM conditional on the one outstanding comment.

**Pull request merged**
